src/main/java/org/embulk/util/json/InternalJsonValuesReader.java

```diff
@@ -0,0 +1,440 @@
+            if (!captures.isEmpty()) {
+                this.builderStack.push(new ObjectBuilder());
+            }
+        } else if (token == JsonToken.END_OBJECT) {
+            if (this.parsingStack.isEmpty() || !this.parsingStack.pop().isObject()) {
+                throw new JsonParseException("END_OBJECT does not match.");
+            }
+        } else if (token == JsonToken.FIELD_NAME) {
+            if (this.parsingStack.isEmpty()) {
+                throw new JsonParseException("FIELD_NAME out of JSON Object.");
+            } else {
+                final ParsingContext context = this.parsingStack.getFirst();
+                if (!context.isObject()) {
+                    throw new JsonParseException("FIELD_NAME in JSON Array.");
+                }
+                context.setPropertyName(this.parser.getCurrentName());
+            }
+        } else if (!token.isScalarValue()) {
+            throw new JsonParseException("Unexpected token in JSON: " + token.toString());
+        }
+
+        if (token.isScalarValue() || token.isStructEnd()) {
+            final JsonValue value;
+            if (token.isScalarValue()) {
+                value = this.getScalarValue(token);
+            } else {  // Structure end
+                if (this.builderStack.isEmpty()) {
+                    value = null;
+                } else {
+                    final StructureBuilder thisBuilder = this.builderStack.pop();
+                    if (token == JsonToken.END_ARRAY && thisBuilder.isObject()) {
+                        throw new JsonParseException("END_ARRAY does not match.");
+                    }
+                    if (token == JsonToken.END_OBJECT && thisBuilder.isArray()) {
+                        throw new JsonParseException("END_OBJECT does not match.");
+                    }
+                    value = thisBuilder.build();
+                }
+            }
+
+            if (value != null) {
+                final JsonPointerTree thisPointer = this.pointerStack.peekFirst();
+                for (final int capture : thisPointer.captures()) {
+                    this.values[capture] = value;
+                }
+
+                if (!this.builderStack.isEmpty()) {
+                    final StructureBuilder parentBuilder = this.builderStack.getFirst();
+                    if (parentBuilder.isArray()) {
+                        parentBuilder.add(value);
+                    } else {  // Object
+                        final ParsingContext context = this.parsingStack.peekFirst();
+                        // If context == null, it's the end of JSON to be parsed while a JSON Pointer "/" is specified.
+                        if (context != null) {
+                            if (!context.isObject()) {
+                                throw new JsonParseException("END_OBJECT does not match.");
+                            } else {
+                                parentBuilder.put(context.getPropertyName(), value);
```

Comment:
Do we need to check duplicate keys like `{"foo":123,"foo":456}`?

Reply:
Skipped the duplication check intentionally. It's a responsibility of the data owner.